### ISSUE TITLE
Example of :setup-children from the manual does not work

### ISSUE THREAD

**Issue by karthink** [magit/transient#343]
In the manual for v0.7.7 of Transient, the following function is provided as an example of how to use `:setup-children`:

```elisp
(transient-define-prefix my-finder-by-keyword ()
  "Select a keyword and list matching packages."
  ;; The real `finder-by-keyword' is more convenient
  ;; of course, but that is not the point here.
  [:class transient-columns
   :setup-children
   (lambda (_)
     (transient-parse-suffixes
      'my-finder-by-keyword
      (let ((char (1- ?A)))
        (mapcar                  ; a list ...
         (lambda (partition)
           (vconcat              ; of group vectors ...
            (mapcar (lambda (elt)
                      (let ((keyword (symbol-name (car elt))))
                                        ; ... where each suffix is a list
                        (list (format "%c" (cl-incf char))
                              keyword
                              (lambda ()
                                (interactive)
                                (finder-list-matches keyword)))))
                    partition)))
         (seq-partition finder-known-keywords 7)))))])
```

When I evaluate this in a lisp-interaction buffer, `my-finder-by-keyword` does not run.  I get the following backtrace:

<details>
<summary>Backtrace</summary>

```elisp
Debugger entered--Lisp error: (void-function closure)
  (transient--exit-and-debug error (void-function closure))
  (closure ((keyword . "abbrev")) nil (interactive) (finder-list-matches keyword))
  (defalias 'transient:my-finder-by-keyword::13106 (closure ((keyword . "abbrev")) nil (interactive) (finder-list-matches keyword)))
  (prog1 'transient:my-finder-by-keyword::13106 (put 'transient:my-finder-by-keyword::13106 'interactive-only t) (put 'transient:my-finder-by-keyword::13106 'completion-predicate #'transient--suffix-only) (defalias 'transient:my-finder-by-keyword::13106 (closure ((keyword . "abbrev")) nil (interactive) (finder-list-matches keyword))))
  (list :key "A" :description "abbrev" :command (prog1 'transient:my-finder-by-keyword::13106 (put 'transient:my-finder-by-keyword::13106 'interactive-only t) (put 'transient:my-finder-by-keyword::13106 'completion-predicate #'transient--suffix-only) (defalias 'transient:my-finder-by-keyword::13106 (closure ((keyword . "abbrev")) nil (interactive) (finder-list-matches keyword)))))
  (list 1 'transient-suffix (list :key "A" :description "abbrev" :command (prog1 'transient:my-finder-by-keyword::13106 (put 'transient:my-finder-by-keyword::13106 'interactive-only t) (put 'transient:my-finder-by-keyword::13106 'completion-predicate #'transient--suffix-only) (defalias 'transient:my-finder-by-keyword::13106 (closure ((keyword . "abbrev")) nil (interactive) (finder-list-matches keyword))))))
  (list (list 1 'transient-suffix (list :key "A" :description "abbrev" :command (prog1 'transient:my-finder-by-keyword::13106 (put 'transient:my-finder-by-keyword::13106 'interactive-only t) (put 'transient:my-finder-by-keyword::13106 'completion-predicate #'transient--suffix-only) (defalias 'transient:my-finder-by-keyword::13106 (closure ((keyword . "abbrev")) nil (interactive) (finder-list-matches keyword)))))) (list 1 'transient-suffix (list :key "B" :description "bib" :command (prog1 'transient:my-finder-by-keyword::13107 (put 'transient:my-finder-by-keyword::13107 'interactive-only t) (put 'transient:my-finder-by-keyword::13107 'completion-predicate #'transient--suffix-only) (defalias 'transient:my-finder-by-keyword::13107 (closure ((keyword . "bib")) nil (interactive) (finder-list-matches keyword)))))) (list 1 'transient-suffix (list :key "C" :description "c" :command (prog1 'transient:my-finder-by-keyword::13108 (put 'transient:my-finder-by-keyword::13108 'interactive-only t) (put 'transient:my-finder-by-keyword::13108 'completion-predicate #'transient--suffix-only) (defalias 'transient:my-finder-by-keyword::13108 (closure ((keyword . "c")) nil (interactive) (finder-list-matches keyword)))))) (list 1 'transient-suffix (list :key "D" :description "calendar" :command (prog1 'transient:my-finder-by-keyword::13109 (put 'transient:my-finder-by-keyword::13109 'interactive-only t) (put 'transient:my-finder-by-keyword::13109 'completion-predicate #'transient--suffix-only) (defalias 'transient:my-finder-by-keyword::13109 (closure ((keyword . "calendar")) nil (interactive) (finder-list-matches keyword)))))) (list 1 'transient-suffix (list :key "E" :description "comm" :command (prog1 'transient:my-finder-by-keyword::13110 (put 'transient:my-finder-by-keyword::13110 'interactive-only t) (put 'transient:my-finder-by-keyword::13110 'completion-predicate #'transient--suffix-only) (defalias 'transient:my-finder-by-keyword::13110 (closure ((keyword . "comm")) nil (interactive) (finder-list-matches keyword)))))) (list 1 'transient-suffix (list :key "F" :description "convenience" :command (prog1 'transient:my-finder-by-keyword::13111 (put 'transient:my-finder-by-keyword::13111 'interactive-only t) (put 'transient:my-finder-by-keyword::13111 'completion-predicate #'transient--suffix-only) (defalias 'transient:my-finder-by-keyword::13111 (closure ((keyword . "convenience")) nil (interactive) (finder-list-matches keyword)))))) (list 1 'transient-suffix (list :key "G" :description "data" :command (prog1 'transient:my-finder-by-keyword::13112 (put 'transient:my-finder-by-keyword::13112 'interactive-only t) (put 'transient:my-finder-by-keyword::13112 'completion-predicate #'transient--suffix-only) (defalias 'transient:my-finder-by-keyword::13112 (closure ((keyword . "data")) nil (interactive) (finder-list-matches keyword)))))))
  (vector 1 'transient-column nil (list (list 1 'transient-suffix (list :key "A" :description "abbrev" :command (prog1 'transient:my-finder-by-keyword::13106 (put 'transient:my-finder-by-keyword::13106 'interactive-only t) (put 'transient:my-finder-by-keyword::13106 'completion-predicate #'transient--suffix-only) (defalias 'transient:my-finder-by-keyword::13106 (closure (...) nil (interactive) (finder-list-matches keyword)))))) (list 1 'transient-suffix (list :key "B" :description "bib" :command (prog1 'transient:my-finder-by-keyword::13107 (put 'transient:my-finder-by-keyword::13107 'interactive-only t) (put 'transient:my-finder-by-keyword::13107 'completion-predicate #'transient--suffix-only) (defalias 'transient:my-finder-by-keyword::13107 (closure (...) nil (interactive) (finder-list-matches keyword)))))) (list 1 'transient-suffix (list :key "C" :description "c" :command (prog1 'transient:my-finder-by-keyword::13108 (put 'transient:my-finder-by-keyword::13108 'interactive-only t) (put 'transient:my-finder-by-keyword::13108 'completion-predicate #'transient--suffix-only) (defalias 'transient:my-finder-by-keyword::13108 (closure (...) nil (interactive) (finder-list-matches keyword)))))) (list 1 'transient-suffix (list :key "D" :description "calendar" :command (prog1 'transient:my-finder-by-keyword::13109 (put 'transient:my-finder-by-keyword::13109 'interactive-only t) (put 'transient:my-finder-by-keyword::13109 'completion-predicate #'transient--suffix-only) (defalias 'transient:my-finder-by-keyword::13109 (closure (...) nil (interactive) (finder-list-matches keyword)))))) (list 1 'transient-suffix (list :key "E" :description "comm" :command (prog1 'transient:my-finder-by-keyword::13110 (put 'transient:my-finder-by-keyword::13110 'interactive-only t) (put 'transient:my-finder-by-keyword::13110 'completion-predicate #'transient--suffix-only) (defalias 'transient:my-finder-by-keyword::13110 (closure (...) nil (interactive) (finder-list-matches keyword)))))) (list 1 'transient-suffix (list :key "F" :description "convenience" :command (prog1 'transient:my-finder-by-keyword::13111 (put 'transient:my-finder-by-keyword::13111 'interactive-only t) (put 'transient:my-finder-by-keyword::13111 'completion-predicate #'transient--suffix-only) (defalias 'transient:my-finder-by-keyword::13111 (closure (...) nil (interactive) (finder-list-matches keyword)))))) (list 1 'transient-suffix (list :key "G" :description "data" :command (prog1 'transient:my-finder-by-keyword::13112 (put 'transient:my-finder-by-keyword::13112 'interactive-only t) (put 'transient:my-finder-by-keyword::13112 'completion-predicate #'transient--suffix-only) (defalias 'transient:my-finder-by-keyword::13112 (closure (...) nil (interactive) (finder-list-matches keyword))))))))
  (transient-parse-suffix my-finder-by-keyword [("A" "abbrev" (closure ((keyword . "abbrev")) nil (interactive) (finder-list-matches keyword))) ("B" "bib" (closure ((keyword . "bib")) nil (interactive) (finder-list-matches keyword))) ("C" "c" (closure ((keyword . "c")) nil (interactive) (finder-list-matches keyword))) ("D" "calendar" (closure ((keyword . "calendar")) nil (interactive) (finder-list-matches keyword))) ("E" "comm" (closure ((keyword . "comm")) nil (interactive) (finder-list-matches keyword))) ("F" "convenience" (closure ((keyword . "convenience")) nil (interactive) (finder-list-matches keyword))) ("G" "data" (closure ((keyword . "data")) nil (interactive) (finder-list-matches keyword)))])
  (apply transient-parse-suffix (my-finder-by-keyword [("A" "abbrev" (closure ((keyword . "abbrev")) nil (interactive) (finder-list-matches keyword))) ("B" "bib" (closure ((keyword . "bib")) nil (interactive) (finder-list-matches keyword))) ("C" "c" (closure ((keyword . "c")) nil (interactive) (finder-list-matches keyword))) ("D" "calendar" (closure ((keyword . "calendar")) nil (interactive) (finder-list-matches keyword))) ("E" "comm" (closure ((keyword . "comm")) nil (interactive) (finder-list-matches keyword))) ("F" "convenience" (closure ((keyword . "convenience")) nil (interactive) (finder-list-matches keyword))) ("G" "data" (closure ((keyword . "data")) nil (interactive) (finder-list-matches keyword)))]))
  (#f(compiled-function (&rest args2) #<bytecode -0xae2f2c9ed0ee63f>) [("A" "abbrev" (closure ((keyword . "abbrev")) nil (interactive) (finder-list-matches keyword))) ("B" "bib" (closure ((keyword . "bib")) nil (interactive) (finder-list-matches keyword))) ("C" "c" (closure ((keyword . "c")) nil (interactive) (finder-list-matches keyword))) ("D" "calendar" (closure ((keyword . "calendar")) nil (interactive) (finder-list-matches keyword))) ("E" "comm" (closure ((keyword . "comm")) nil (interactive) (finder-list-matches keyword))) ("F" "convenience" (closure ((keyword . "convenience")) nil (interactive) (finder-list-matches keyword))) ("G" "data" (closure ((keyword . "data")) nil (interactive) (finder-list-matches keyword)))])
  (transient-parse-suffixes my-finder-by-keyword ([("A" "abbrev" (closure ((keyword . "abbrev")) nil (interactive) (finder-list-matches keyword))) ("B" "bib" (closure ((keyword . "bib")) nil (interactive) (finder-list-matches keyword))) ("C" "c" (closure ((keyword . "c")) nil (interactive) (finder-list-matches keyword))) ("D" "calendar" (closure ((keyword . "calendar")) nil (interactive) (finder-list-matches keyword))) ("E" "comm" (closure ((keyword . "comm")) nil (interactive) (finder-list-matches keyword))) ("F" "convenience" (closure ((keyword . "convenience")) nil (interactive) (finder-list-matches keyword))) ("G" "data" (closure ((keyword . "data")) nil (interactive) (finder-list-matches keyword)))] [("H" "docs" (closure ((keyword . "docs")) nil (interactive) (finder-list-matches keyword))) ("I" "emulations" (closure ((keyword . "emulations")) nil (interactive) (finder-list-matches keyword))) ("J" "extensions" (closure ((keyword . "extensions")) nil (interactive) (finder-list-matches keyword))) ("K" "faces" (closure ((keyword . "faces")) nil (interactive) (finder-list-matches keyword))) ("L" "files" (closure ((keyword . "files")) nil (interactive) (finder-list-matches keyword))) ("M" "frames" (closure ((keyword . "frames")) nil (interactive) (finder-list-matches keyword))) ("N" "games" (closure ((keyword . "games")) nil (interactive) (finder-list-matches keyword)))] [("O" "hardware" (closure ((keyword . "hardware")) nil (interactive) (finder-list-matches keyword))) ("P" "help" (closure ((keyword . "help")) nil (interactive) (finder-list-matches keyword))) ("Q" "hypermedia" (closure ((keyword . "hypermedia")) nil (interactive) (finder-list-matches keyword))) ("R" "i18n" (closure ((keyword . "i18n")) nil (interactive) (finder-list-matches keyword))) ("S" "internal" (closure ((keyword . "internal")) nil (interactive) (finder-list-matches keyword))) ("T" "languages" (closure ((keyword . "languages")) nil (interactive) (finder-list-matches keyword))) ("U" "lisp" (closure ((keyword . "lisp")) nil (interactive) (finder-list-matches keyword)))] [("V" "local" (closure ((keyword . "local")) nil (interactive) (finder-list-matches keyword))) ("W" "maint" (closure ((keyword . "maint")) nil (interactive) (finder-list-matches keyword))) ("X" "mail" (closure ((keyword . "mail")) nil (interactive) (finder-list-matches keyword))) ("Y" "matching" (closure ((keyword . "matching")) nil (interactive) (finder-list-matches keyword))) ("Z" "mouse" (closure ((keyword . "mouse")) nil (interactive) (finder-list-matches keyword))) ("[" "multimedia" (closure ((keyword . "multimedia")) nil (interactive) (finder-list-matches keyword))) ("\\" "news" (closure ((keyword . "news")) nil (interactive) (finder-list-matches keyword)))] [("]" "outlines" (closure ((keyword . "outlines")) nil (interactive) (finder-list-matches keyword))) ("^" "processes" (closure ((keyword . "processes")) nil (interactive) (finder-list-matches keyword))) ("_" "terminals" (closure ((keyword . "terminals")) nil (interactive) (finder-list-matches keyword))) ("`" "tex" (closure ((keyword . "tex")) nil (interactive) (finder-list-matches keyword))) ("a" "tools" (closure ((keyword . "tools")) nil (interactive) (finder-list-matches keyword))) ("b" "unix" (closure ((keyword . "unix")) nil (interactive) (finder-list-matches keyword))) ("c" "vc" (closure ((keyword . "vc")) nil (interactive) (finder-list-matches keyword)))] [("d" "wp" (closure ((keyword . "wp")) nil (interactive) (finder-list-matches keyword)))]))
  ((closure (t) (_) (transient-parse-suffixes 'my-finder-by-keyword (let ((char (1- 65))) (mapcar #'(lambda (partition) (vconcat ...)) (seq-partition finder-known-keywords 7))))) nil)
  (transient-setup-children #<transient-columns transient-columns-7867cb8> nil)
  (transient--init-group nil [1 transient-columns (:setup-children (closure (t) (_) (transient-parse-suffixes 'my-finder-by-keyword (let ((char ...)) (mapcar #'... (seq-partition finder-known-keywords 7)))))) nil] nil)
  (transient--init-child nil [1 transient-columns (:setup-children (closure (t) (_) (transient-parse-suffixes 'my-finder-by-keyword (let ((char ...)) (mapcar #'... (seq-partition finder-known-keywords 7)))))) nil] nil)
  (#f(compiled-function (c) #<bytecode -0x142f447a72b3159a>) [1 transient-columns (:setup-children (closure (t) (_) (transient-parse-suffixes 'my-finder-by-keyword (let ((char ...)) (mapcar #'... (seq-partition finder-known-keywords 7)))))) nil])
  (cl-mapcan #f(compiled-function (c) #<bytecode -0x142f447a72b3159a>) ([1 transient-columns (:setup-children (closure (t) (_) (transient-parse-suffixes 'my-finder-by-keyword (let (...) (mapcar ... ...))))) nil] [1 transient-columns (:hide #f(compiled-function () #<bytecode 0x173b9463840ed272>)) ([1 transient-column (:description "Value commands") ((1 transient-suffix (:key "C-x s  " :description "Set" :command transient-set)) (1 transient-suffix (:key "C-x C-s" :description "Save" :command transient-save)) (1 transient-suffix (:key "C-x C-k" :description "Reset" :command transient-reset)) (1 transient-suffix (:key "C-x p  " :description "Previous value" :command transient-history-prev)) (1 transient-suffix (:key "C-x n  " :description "Next value" :command transient-history-next)))] [1 transient-column (:description "Sticky commands") ((1 transient-suffix (:key "C-g" :description "Quit prefix or transient" :command transient-quit-one)) (1 transient-suffix (:key "C-q" :description "Quit transient stack" :command transient-quit-all)) (1 transient-suffix (:key "C-z" :description "Suspend transient stack" :command transient-suspend)))] [1 transient-column (:description "Customize") ((1 transient-suffix (:key "C-x t" :command transient-toggle-common :description #f(compiled-function () #<bytecode 0x1e0d3d51f163efb5>))) (1 transient-suffix (:key "C-x l" :description "Show/hide suffixes" :command transient-set-level)) (1 transient-suffix (:key "C-x a" :command transient-toggle-level-limit)))])]))
  (transient--init-suffixes my-finder-by-keyword)
  (transient--init-objects my-finder-by-keyword nil nil)
  (transient-setup my-finder-by-keyword)
  (my-finder-by-keyword)
```

</details>

It appears that because `lexical-binding` is set, the lambdas evaluate to closures, and Transient is not expecting that.  I had to quote the lambdas instead:

```elisp
(transient-define-prefix my-finder-by-keyword ()
  "Select a keyword and list matching packages."
  ;; The real `finder-by-keyword' is more convenient
  ;; of course, but that is not the point here.
  [:class transient-columns
   :setup-children
   (lambda (_)
     (transient-parse-suffixes
      'my-finder-by-keyword
      (let ((char (1- ?A)))
        (mapcar                  ; a list ...
         (lambda (partition)
           (vconcat              ; of group vectors ...
            (mapcar (lambda (elt)
                      (let ((keyword (symbol-name (car elt))))
                                        ; ... where each suffix is a list
                        (list (format "%c" (cl-incf char))
                              keyword
                              '(lambda ()
                                (interactive)
                                (finder-list-matches keyword)))))
                    partition)))
         (seq-partition finder-known-keywords 7)))))])
```

It might be worth updating the example in the manual.  My apologies if this is already fixed in a later version of transient, please close this issue in that case.

----------

On a related note, why do the lambdas have to be quoted/passed in as data instead of as functions?

**Comment by tarsius** [comment]
I wasn't able to reproduce this with v0.8.1 (well actually with what will be released as v0.8.2 in two days).

**Comment by tarsius** [comment]
> On a related note, why do the lambdas have to be quoted/passed in as data instead of as functions?

It's ... complicated. Basically because we *prevent* it from being evaluated by placing it `[...here...]`, we later have to evaluate it explicitly. In doing so we have to make sure we don't end up doing it twice.
